### PR TITLE
(DONE) Registration of arguments of plugins, add invisible mode (No splashscreen!) and "--debug" mode as argument

### PR DESCRIPTION
**Pull request by thopiekar**
_Features:_
 * Adds an early argparser to cura_app and passes it through the whole application
 * Now with "--debug" you can get all messages into std*. Early errors are this way visible, because we skip the CrashHandler also.

_Depends on:_
 * https://github.com/Ultimaker/Uranium/pull/316

_Includes:_
* https://github.com/Ultimaker/Cura/pull/2939

_Needed for:_
 * Individual arguments in the InterfacePlugin (https://github.com/thopiekar/CuraInterfacePlugin)

_In combination with:_
 * https://github.com/thopiekar/CuraInterfacePlugin

### CODE REVIEW

**Comment by thopiekar** [comment]
The difference between with and without "--debug":
```
```
PS C:\Program Files\Ultimaker Cura 3.2> .\CuraCLI.exe --debug
Traceback (most recent call last):
  File "<frozen importlib._bootstrap>", line 969, in _find_and_load
  File "<frozen importlib._bootstrap>", line 958, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 664, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 634, in _load_backward_compatible
  File "__startup__.py", line 12, in <module>
  File "<frozen importlib._bootstrap>", line 969, in _find_and_load
  File "<frozen importlib._bootstrap>", line 958, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 664, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 634, in _load_backward_compatible
  File "Console.py", line 24, in <module>
  File "C:\cura_buildsystem\binaries\package\inst\bin\cura_app.py", line 93, in <module>
    if not cura.CuraApplication.CuraApplication.preStartUp(parser = parser, parsed_command_line = known_args):
  File "C:\cura_buildsystem\binaries\package\inst\lib\python3.5\site-packages\cura\CuraApplication.py", line 634, in preStartUp
    CuraApplication.addCommandLineOptions(parser)
  File "C:\cura_buildsystem\binaries\package\inst\lib\python3.5\site-packages\cura\CuraApplication.py", line 573, in addCommandLineOptions
    super().addCommandLineOptions(parser)
  File "C:\cura_buildsystem\binaries\package\inst\lib\python3.5\site-packages\UM\Qt\QtApplication.py", line 282, in addCommandLineOptions
    super().addCommandLineOptions(parser)
  File "C:\cura_buildsystem\binaries\package\inst\lib\python3.5\site-packages\UM\Application.py", line 363, in addCommandLineOptions
    if "debug" not in cls._parsed_command_line:
AttributeError: type object 'CuraApplication' has no attribute '_parsed_command_line'
PS C:\Program Files\Ultimaker Cura 3.2> .\CuraCLI.exe
QWidget: Must construct a QApplication before a QWidget
```
```

**Comment by thopiekar** [comment]
Currently when running `CuraCLI.exe --debug --invisible`, the window pops up after the engine has been created. Have to find a different way to keep it hidden..
```
Traceback (most recent call last):
  File "C:\cura_buildsystem\binaries\package\inst\lib\python3.5\site-packages\UM\Qt\Bindings\MainWindow.py", line 201, in _render
    renderer.render()
  File "C:\cura_buildsystem\binaries\package\inst\lib\python3.5\site-packages\UM\Qt\QtRenderer.py", line 148, in render
    render_pass.render()
  File "C:\cura_buildsystem\binaries\package\inst\lib\python3.5\site-packages\UM\View\SelectionPass.py", line 81, in render
    self.release()
  File "C:\cura_buildsystem\binaries\package\inst\lib\python3.5\site-packages\UM\View\RenderPass.py", line 93, in release
    self._fbo.release()
AttributeError: 'NoneType' object has no attribute 'release'
Fatal Python error: Aborted

Thread 0x00001758 (most recent call first):
  File "threading.py", line 293 in wait
  File "threading.py", line 424 in acquire
  File "C:\cura_buildsystem\binaries\package\inst\lib\python3.5\site-packages\UM\JobQueue.py", line 86 in _nextJob
  File "C:\cura_buildsystem\binaries\package\inst\lib\python3.5\site-packages\UM\JobQueue.py", line 116 in run
  File "threading.py", line 914 in _bootstrap_inner
  File "threading.py", line 882 in _bootstrap

Thread 0x00005010 (most recent call first):
  File "threading.py", line 293 in wait
  File "threading.py", line 424 in acquire
  File "C:\cura_buildsystem\binaries\package\inst\lib\python3.5\site-packages\UM\JobQueue.py", line 86 in _nextJob
  File "C:\cura_buildsystem\binaries\package\inst\lib\python3.5\site-packages\UM\JobQueue.py", line 116 in run
  File "threading.py", line 914 in _bootstrap_inner
  File "threading.py", line 882 in _bootstrap

Thread 0x00001260 (most recent call first):
  File "threading.py", line 293 in wait
  File "threading.py", line 424 in acquire
  File "C:\cura_buildsystem\binaries\package\inst\lib\python3.5\site-packages\UM\JobQueue.py", line 86 in _nextJob
  File "C:\cura_buildsystem\binaries\package\inst\lib\python3.5\site-packages\UM\JobQueue.py", line 116 in run
  File "threading.py", line 914 in _bootstrap_inner
  File "threading.py", line 882 in _bootstrap

Thread 0x0000444c (most recent call first):
  File "threading.py", line 293 in wait
  File "threading.py", line 424 in acquire
  File "C:\cura_buildsystem\binaries\package\inst\lib\python3.5\site-packages\UM\JobQueue.py", line 86 in _nextJob
  File "C:\cura_buildsystem\binaries\package\inst\lib\python3.5\site-packages\UM\JobQueue.py", line 116 in run
  File "threading.py", line 914 in _bootstrap_inner
  File "threading.py", line 882 in _bootstrap

Thread 0x0000083c (most recent call first):
  File "threading.py", line 297 in wait
  File "zeroconf.py", line 1569 in wait
  File "zeroconf.py", line 1153 in run
  File "threading.py", line 914 in _bootstrap_inner
  File "threading.py", line 882 in _bootstrap

Thread 0x00005208 (most recent call first):
  File "threading.py", line 297 in wait
  File "zeroconf.py", line 1569 in wait
  File "zeroconf.py", line 1017 in run
  File "threading.py", line 914 in _bootstrap_inner
  File "threading.py", line 882 in _bootstrap

Thread 0x000024c0 (most recent call first):
  File "zeroconf.py", line 946 in run
  File "threading.py", line 914 in _bootstrap_inner
  File "threading.py", line 882 in _bootstrap

Thread 0x00004474 (most recent call first):
  File "C:\Program Files\Ultimaker Cura 3.2\plugins\RemovableDriveOutputDevice\RemovableDrivePlugin.py", line 61 in _updateThread
  File "threading.py", line 862 in run
  File "threading.py", line 914 in _bootstrap_inner
  File "threading.py", line 882 in _bootstrap

Thread 0x00002ec0 (most recent call first):
  File "threading.py", line 297 in wait
  File "zeroconf.py", line 1569 in wait
  File "zeroconf.py", line 1153 in run
  File "threading.py", line 914 in _bootstrap_inner
  File "threading.py", line 882 in _bootstrap

Thread 0x0000155c (most recent call first):
  File "threading.py", line 297 in wait
  File "zeroconf.py", line 1569 in wait
  File "zeroconf.py", line 1017 in run
  File "threading.py", line 914 in _bootstrap_inner
  File "threading.py", line 882 in _bootstrap

Thread 0x000000b8 (most recent call first):
  File "zeroconf.py", line 946 in run
  File "threading.py", line 914 in _bootstrap_inner
  File "threading.py", line 882 in _bootstrap

Current thread 0x00003a74 (most recent call first):
  File "C:\cura_buildsystem\binaries\package\inst\lib\python3.5\site-packages\cura\CuraApplication.py", line 744 in run
  File "C:\cura_buildsystem\binaries\package\inst\bin\cura_app.py", line 98 in <module>
  File "Console.py", line 24 in <module>
  File "<frozen importlib._bootstrap>", line 634 in _load_backward_compatible
  File "<frozen importlib._bootstrap>", line 664 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 958 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 969 in _find_and_load
  File "__startup__.py", line 12 in <module>
  File "<frozen importlib._bootstrap>", line 634 in _load_backward_compatible
  File "<frozen importlib._bootstrap>", line 664 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 958 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 969 in _find_and_load
```

**Comment by thopiekar** [comment]
Not 100%ly sure, but I think I finished my work here. That is the prove of my working concept:
(Yes, every plugin can not register it's own arguments.)
image

**Comment by Appesteijn** [comment]
CURA-4694

File: cura/CuraApplication.py [hunk 1]
```diff
@@ -716,12 +741,11 @@ def run(self):
         self.setMainQml(Resources.getPath(self.ResourceTypes.QmlFiles, "Cura.qml"))
         self._qml_import_paths.append(Resources.getPath(self.ResourceTypes.QmlFiles))
 
-        run_headless = self.getCommandLineOption("headless", False)
-        if not run_headless:
+        run_without_gui = self.getCommandLineOption("headless", False) or self.getCommandLineOption("invisible", False)
+        if not run_without_gui:
             self.initializeEngine()
-            controller.setActiveStage("PrepareStage")
```
Comment (ChrisTerBeke) [inline cura/CuraApplication.py]: Why did this line get removed? It's needed to set Cura to the correct stage on start.

Comment (thopiekar) [reply cura/CuraApplication.py]: Has been moved to Uranium, but you probably already noticed.

Comment (ChrisTerBeke) [reply cura/CuraApplication.py]: I have. I moved it back though as Uranium doesn't know about a "PrepareStage", that's Cura specific.

**Pull request merged**